File: 025_reverse_nodes_in_k_group/nonrecursion2/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy_head = ListNode(-1)
        dummy_head.next = head
        pre, end = dummy_head, dummy_head

        while end.next:
            for i in range(k):
                end = end.next
                if not end: return dummy_head.next
            start = pre.next
            next_node = end.next
            pre.next = self.reverse(start, next_node)
            start.next = next_node
            pre = start
            end = pre

        return dummy_head.next

    def reverse(self, head: ListNode, tail: ListNode) -> ListNode:
        pre = None
        while head != tail:
            next_node = head.next
            head.next, pre, head = pre, head, next_node

        return pre


if __name__ == '__main__':
    head = ListNode(1)
    cur = head
    for i in range(2, 10):
        cur.next = ListNode(i)
        cur = cur.next

    solution = Solution()
    cur = solution.reverseKGroup(head, 2)
    while cur:
        print(cur.val, end = ' ')
        cur = cur.next

File: 025_reverse_nodes_in_k_group/nonrecursion2/Solution.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
    }
}

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummyHead = new ListNode(-1);
        dummyHead.next = head;
        ListNode pre = dummyHead, end = dummyHead;
        ListNode start, next;

        while (end.next != null) {
            for (int i = 0; i < k; i++) {
                end = end.next;
                if (end == null) return dummyHead.next;
            }
            start = pre.next;
            next = end.next; 
            pre.next = reverse(start, next); 
            start.next = next;
            pre = start;
            end = pre;
        }

        return dummyHead.next;
    }

    public ListNode reverse(ListNode head, ListNode tail) {
        ListNode pre = null, next;

        while (head != tail) {
            next = head.next;
            head.next = pre;
            pre = head;
            head = next;
        }

        return pre;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        ListNode cur = head;
        for (int i = 2; i < 10; i++) {
            cur.next = new ListNode(i);
            cur = cur.next;
        }

        Solution solution = new Solution();
        cur = solution.reverseKGroup(head, 2);
        while (cur != null) {
            System.out.print(cur.val + " ");
            cur = cur.next;
        }
    }
}
